ts/reviewer/image_occlusion.ts

```diff
@@ -0,0 +1,79 @@
+// Copyright: Ankitects Pty Ltd and contributors
+// License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
+
+export function setupImageCloze() {
+    const canvas: HTMLCanvasElement = document.createElement("CANVAS") as HTMLCanvasElement;
+    canvas.id = "image-occlusion-canvas";
+    canvas.style.backgroundSize = "100% 100%";
+    canvas.style.maxWidth = "100%";
+    canvas.style.maxHeight = "90vh";
+
+    const ctx: CanvasRenderingContext2D = canvas.getContext("2d")!;
+    const imageFieldElem = document.getElementById("img") as HTMLImageElement;
+    imageFieldElem.style.display = "none";
+
+    const img: HTMLImageElement = new Image();
+    img.src = imageFieldElem.src;
+
+    img.onload = () => {
+        canvas.width = img.width;
+        canvas.height = img.height;
+        ctx.drawImage(img, 0, 0, img.width, img.height);
+        drawShapes(ctx);
+    };
+
+    document.getElementById("qa")!.appendChild(canvas);
+}
+
+function drawShapes(ctx: CanvasRenderingContext2D) {
+    const activeCloze = document.querySelectorAll(".cloze");
+    const inActiveCloze = document.querySelectorAll(".cloze-inactive");
+
+    for (let clz of activeCloze) {
+        let cloze = (<HTMLDivElement>clz);
+        const shape = cloze.dataset.shape!;
+        const fill = cloze.dataset.quesmaskcolor!;
+        draw(ctx, cloze, shape, fill);
+    }
+
+    for (let clz of inActiveCloze) {
+        let cloze = (<HTMLDivElement>clz);
+        const shape = cloze.dataset.shape!;
+        const fill = cloze.dataset.fill!;
+        draw(ctx, cloze, shape, fill);
+    }
+}
+
+function draw(ctx: CanvasRenderingContext2D, cloze: HTMLDivElement, shape: string, color: string) {
+    ctx.fillStyle = color;
+
+    const post_left = parseFloat(cloze.dataset.left!);
```

Comment:
Food for thought: Would we want to check whether the values here are actually valid? Or in other words: Say an accidental `e` character finds its way into one of the float-type fields, which would persist into the rendered cloze as we do not check the values during their processing in `get_image_cloze_data`, what would the behavior of the template be? Or in other words, would we want to specifically account for invalid data and communicate this to users somehow?

Given the right precautions implemented into Anki clients, hopefully users won't run into invalid data too often, but we should probably check how card rendering would behave in cases like this and consider whether we want to introduce error handling here.

Reply:
The `{{c1::...` values are passed to `image_occlusion.rs` and then it transformed into `data-*`. The check have implemented for only empty but above issues needs to address in that implementation. I will update it.

> A couple of thoughts regarding `setupImageCloze()`:
> 
> * How were you thinking of encoding the "Hide all" vs "Hide one" behavior? I was thinking that this is something that we could specify as an argument to `setupImageCloze()`, e.g. a `hideAll` bool flag. To make the interface extensible and expressive, we could group all customizable properties in a single-argument object, i.e.:

I was thinking about about adding the `hideall` value in cloze itself but above idea is better. I will implement using above idea. Thanks

Reply:
Now hideinactive option added to cloze itself. It is more customizable.
Lets say we have group cloze we want to hide one shape for some reason instead of removing from group cloze then changing hideinactive to true will not show during review.
Also the value will be used for hide one guess one note types.